File: Data Structures & Algorithms/duplicate-integer/submission-0.cpp
class Solution {
public:
    bool hasDuplicate(vector<int>& nums) {
        unordered_map<int,int> counts;

        for(int num : nums) {
            if(counts.contains(num)) return true;
            counts[num]=1;
        }

        return false;
    }
};

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-0.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        dp = {}
        def check(n, s1=0, s2=0, c=0):
            if c==len(n):
                if s1==s2: return True
                return False

            if (s1,s2,c) in dp: return dp[(s1,s2,c)]

            dp[(s1,s2,c)] = check(n,s1+n[c],s2,c+1) or check(n,s1,s2+n[c],c+1)
            return dp[(s1,s2,c)]

        return check(nums)